### ISSUE TITLE
Why is specifying a policy reform so complicated?

### ISSUE THREAD

**Issue by martinholmer** [PSLmodels/Tax-Calculator#2291]
The way Tax-Calculator users are asked to specify a policy reform has remained the same since the beginning of the project.  Meanwhile, there have been several important ease-of-use enhancements.  Several years ago, a command-line interface to Tax-Calculator was introduced so that no computer programming was required.  And the new Tax-Brain interface to Tax-Calculator and associated USA tax models makes it much easier for Python programmers to use Tax-Calculator.  Given these improvements, the arcane syntax of specifying policy reforms presents new users with an unnecessary challenge in using Tax-Calculator.

In this comment, I describe a simple tax reform in words, and then show how this reform is specified using the current syntax.  After pointing out the  complexities (and inconsistencies) of the current syntax, I show how this reform could be specified in an easier way.

Consider a two-provision reform: (1) raise the FICA maximum taxable earnings to $300,000 in 2020 and leave it at that level for twelve years before resuming indexing; (2) raise the standard deduction in 2021 and continue to index it in subsequent years.

Using the current version of Tax-Calculator, this reform would be specified using the following Python dictionary:
```
reform = {
   2020: {
       '_SS_Earnings_c': [300000],
       '_SS_Earnings_c_cpi': False
   },
   2021: {
       '_STD': [[14000, 28000, 14000, 21000, 28000]]
   }
}
```

When first encountering this syntax, new users probably have the same reactions I had years ago when I started using Tax-Calculator.  Why the need for the leading underscore character at the start of each parameter name?  Why the need for brackets around the value of the maximum taxable earnings when its value is simply a number?  (After all, the boolean value of `_SS_Earnings_c_cpi` doesn't need brackets.)  Why the need for **two** sets of brackets to enclose the vector-valued `_STD` parameter values?  And finally, why is there a `_cpi` suffix on the end of the `_SS_Earnings_c` parameter (to turn off indexing) when that parameter is indexed by wages not prices?

It is not productive to speculate about how this reform syntax came to be.  It is far more productive to think about how it could be improved.

I think ease of use would be improved if users could specify this reform using the following Python dictionary:
```
reform = {
   2020: {
       'SS_Earnings_c': 300000,
       'SS_Earnings_c-indexed': False
   },
   2021: {
       'STD': [14000, 28000, 14000, 21000, 28000]
   }
}
```

I'm not the only one to think the current syntax could be improved.  For example, @codykallen initially created his own policy parameter logic in the Business-Taxation repository, and his logic had no leading underscores and no brackets around scalar parameters.  And @hdoupe has recently developed a parameter-handling capability in the ParamTools repository that does not require parameter names with leading underscores or brackets around scalar parameters.

I think the kind of syntax change suggested above would be a major ease-of-use enhancement to Tax-Calculator.

@MattHJensen @andersonfrailey

**Comment by hdoupe** [comment]
@martinholmer This enhancement looks very nice. I have a couple questions/comments about the enhancement in #2292 and the Tax-Calculator parameters structure in general. I understand that Tax-Calculator may not be interested in the ideas/critiques below, but I wanted to list them while this issue is under consideration.

1. I noticed that `-indexed` is going to be used instead of `_cpi`. So, `SS_Earnings_c_cpi` becomes `SS_Earnings_c-indexed`. To me, it seems like this should be `SS_Earnings_c_indexed`.
2. One longstanding gripe that I have with Tax-Calculator parameters is that there are two parameter formats:
   - JSON based is `parameter: year: value`
   - while Dict based is `year: parameter: value`
   
    The JSON based format is converted using `Calculator.read_json_param_objects` to the Python dict based format which is fed into `Policy.implement_reform` (or some equivalent method).

    I found these two formats very confusing when I first began working with Tax-Calculator. Now, I understand them, but I still find it annoying to have to convert one format to another.

3. When I built ParamTools, I made a couple changes to the parameter attributes. I see that Tax-Calculator has since incorporated some similar changes.
    - I changed "long_name" to "title" because what does "long_name" even mean? Is there a short name? "title" seemed more appropriate to me.
    - I changed "integer_type", "boolean_type", etc. to "type" which can take on values "str", "bool", "int", "float", or "date". I chose these because they mapped directly to Python data types. Tax-Calculator now uses "value_type" which can take on values "real", "boolean", "integer", or "string". I may bring those names over to ParamTools because they are more intuitive to those who are not as familiar with Python.
    - Instead of "valid_values", ParamTools uses "validators." "validators" is a dict of validator objects. These objects can be "range" validators or "choice" validators. An example validator attribute looks like this:
    ```json
    {
        "validators": {
            "range": {"min": 10, "max": 20},
        }
    }
    ```
    or this:
    
    ```json
    {
        "validators": {
            "choice": {"choices": ["hello", "world"]},
        }
    }
    ```

    The advantage of this design is that new validators can be added very intuitively. For example, ["none of"](https://marshmallow.readthedocs.io/en/3.0/api_reference.html#marshmallow.validate.NoneOf) is a common validator where there are certain values that a parameter cannot take on. ParamTools does not currently support "none of," but it could. There doesn't seem like there's an intuitive way to add validators to the Tax-Calculator inputs format.

4. This is related to 2. Tax-Calculator uses numbers as keys in its adjustment data structures (i.e. reforms or assumption files). However, the JSON specification forces all keys to be strings. Thus, serializing and deserializing these adjustment data structures is very annoying. To demonstrate this, here's the function that COMP uses to deserialize Tax-Calculator adjustment data:
    ```python
    def json_int_key_encode(rename_dict):
        """
        Recursively rename integer value keys if they are casted to strings
        via JSON encoding

        returns: dict with new keys
        """
        if isinstance(rename_dict, dict):
            for k in list(rename_dict.keys()):
                if hasattr(k, "isdigit") and k.isdigit():
                    new_label = int(k)
                else:
                    new_label = k
                rename_dict[new_label] = json_int_key_encode(rename_dict.pop(k))
        return rename_dict
    ````

    If Tax-Calculator used a different approach, this would make it easier to serialize and deserialize this data. This isn't just about serialization and deserialization. I've gotten mixed up several times about whether the year needs to be a number or a string and gotten this error:
    ```
    In [1]: import taxcalc

    In [2]: pol = taxcalc.Policy()

    In [3]: pol.implement_reform({"2020": {"II_em": [5000]}})
    ---------------------------------------------------------------------------
    ValueError                                Traceback (most recent call last)
    <ipython-input-3-1f402572f5db> in <module>()
    ----> 1 pol.implement_reform({"2020": {"II_em": [5000]}})

    ~/anaconda3/lib/python3.6/site-packages/taxcalc/policy.py in implement_reform(self, reform, print_warnings, raise_errors)
        172                 msg = 'ERROR: {} KEY {}'
        173                 details = 'KEY in reform is not an integer calendar year'
    --> 174                 raise ValueError(msg.format(year, details))
        175         # check range of remaining reform_years
        176         first_reform_year = min(reform_years)

    ValueError: ERROR: 2020 KEY KEY in reform is not an integer calendar year
    ```

    I almost always have to look in the method docstring to double check what Tax-Calculator is expecting.


Again, I don't expect Tax-Calculator to be interested in making any or all of these changes. This seemed like an appropriate time and place to list these critiques/enhancement ideas. Issues 2, 3, and 4 were very much on my mind when I wrote ParamTools. I'm happy to discuss my solutions for them if there's interest.

**Comment by martinholmer** [comment]
In #2291 @hdoupe said:
> I noticed that `-indexed` is going to be used instead of `_cpi`. So, `SS_Earnings_c_cpi` becomes `SS_Earnings_c-indexed`. To me, it seems like this should be `SS_Earnings_c_indexed`.

I considered that option, but decided on the hyphenation approach to reduce the risk of confusion in the future if any new parameter name ended in `_indexed`.

**Comment by martinholmer** [comment]
In #2291 @hdoupe said:
> One longstanding gripe that I have with Tax-Calculator parameters is that there are two parameter formats:
>
> JSON based is `parameter: year: value`
> while Dict based is `year: parameter: value`
>
> The JSON based format is converted using `Calculator.read_json_param_objects` to the Python dict based format which is fed into `Policy.implement_reform` (or some equivalent method).
>
> I found these two formats very confusing when I first began working with Tax-Calculator. Now, I understand them, but I still find it annoying to have to convert one format to another.

The dict format goes back to very beginning, so I have no idea why it is "backwards" from the way the TaxBrain web app presents the parameters, which is like the JSON-based formats.  Getting rid of these two approaches would involve major work for either TaxBrain or Tax-Calculator, depending on which format was eliminated.

**Comment by martinholmer** [comment]
In #2291 @hdoupe said:
> I changed "long_name" to "title" because what does "long_name" even mean? Is there a short name? "title" seemed more appropriate to me.

You're absolutely correct about `long_name` being sort of ridiculous.  I will update PR #2292 to use simply `name`.

**Comment by MattHJensen** [comment]
> I changed "long_name" to "title" because what does "long_name" even mean? Is there a short name? "title" seemed more appropriate to me.

I don't have a view on how the attribute is relabeled, but for historical reference, the thinking was that `_cpi_offset` was the "short name" and `Decimal offset ADDED to unchained CPI to get parameter indexing rate` was the "long name". 

(Incidentally, in the quote below, Martin references `_cpi_offset` and similar as the "parameter name")

> Why the need for the leading underscore character at the start of each parameter name?

For historical reference, the leading underscore was meant to indicate that the variable contained all years' values, and no leading underscore meant that it contained only the current year's value(s). So in `policy_current_law.json` you would see _ and in `calcfunctions.py` you would not.

**Comment by martinholmer** [comment]
@MattHJensen said in issue #2291:
> I don't have a view on how the attribute is relabeled, but for historical reference, the thinking was that `_cpi_offset` was the "short name" and `Decimal offset ADDED to unchained CPI to get parameter indexing rate` was the "long name".
>
>(Incidentally, in the quote below, Martin references _cpi_offset and similar as the "parameter name")
>> Why the need for the leading underscore character at the start of each parameter name?

@MattHJensen, Thanks for the insight on this.   Given all this, I think its best to stick with the current "long_name" terminology.

**Comment by martinholmer** [comment]
In #2291 @MattHJensen said:
> For historical reference, the leading underscore was meant to indicate that the variable contained all years' values, and no leading underscore meant that it contained only the current year's value(s). So in `policy_current_law.json` you would see `_` and in `calcfunctions.py` you would not.

OK, that's the first time I've heard or seen that clearly stated.  I think it is a good distinction to make for **developers** working on the code.  And that is **still** the way it is internally in PR #2292.  So that distinction is unchanged in #2292.  But what we're discussing in issue #2291 is not developers' ease of use but rather how easy it is for non-developers to use Tax-Calculator.   Why would we expect non-developers (who have no knowledge of Tax-Calculator internals) to understand the leading underscore?  To me we should be giving user ease-of-use a very high priority.

**Comment by hdoupe** [comment]
One option is to store the values for all years in the parameter's entry in the `_vals` dictionary:

```
In [1]: import taxcalc

In [2]: pol = taxcalc.Policy()

In [3]: pol._vals["_STD"]["value"]
Out[3]: 
[[6100.0, 12200.0, 6100.0, 8950.0, 12200.0],
 [6200.0, 12400.0, 6200.0, 9100.0, 12400.0],
 [6300.0, 12600.0, 6300.0, 9250.0, 12600.0],
 [6300.0, 12600.0, 6300.0, 9300.0, 12600.0],
 [6350.0, 12700.0, 6350.0, 9350.0, 12700.0],
 [12000.0, 24000.0, 12000.0, 18000.0, 24000.0],
 [12268.8, 24537.6, 12268.8, 18403.2, 24537.6],
 [12497.0, 24994.0, 12497.0, 18745.5, 24994.0],
 [12788.18, 25576.36, 12788.18, 19182.27, 25576.36],
 [13081.03, 26162.06, 13081.03, 19621.54, 26162.06],
 [13379.28, 26758.55, 13379.28, 20068.91, 26758.55],
 [13674.96, 27349.92, 13674.96, 20512.44, 27349.92],
 [13963.5, 27927.0, 13963.5, 20945.25, 27927.0],
 [7685.0, 15369.0, 7685.0, 11315.0, 15369.0]]
```

Then, the active values are stored in the instance attribute. That way you don't have to fool with the underscore business.

**Comment by martinholmer** [comment]
@hdoupe said in issue #2291:
> One longstanding gripe that I have with Tax-Calculator parameters is that there are two parameter formats:
>
> JSON based is `parameter: year: value`
> while Dict based is `year: parameter: value`
>
> The JSON based format is converted using `Calculator.read_json_param_objects` to the Python dict based format which is fed into `Policy.implement_reform` (or some equivalent method).
>
> I found these two formats very confusing when I first began working with Tax-Calculator. Now, I understand them, but I still find it annoying to have to convert one format to another.

So, which format should do you think we should consider eliminating?

We could make the JSON format the same as the dict format:  `year : parameter : value`, but it seems we would still need a `Calculator.read_json_param_objects` method to read the JSON files and package them as dictionaries.  If we did this, users would have to do more extensive reformatting of their local JSON reform files, and TaxBrain would have to take user input and generate a very different kind of JSON file to send to Tax-Calculator.  @hdoupe, is that easy for TaxBrain to do?  T.J. suggested the current JSON format because it was easy for TaxBrain to translate the screen inputs into a JSON file that was formatted `parameter : year : value` (because that's the way the parameters are shown on the TaxBrain input screen).  Are you willing and able to make the changes to TaxBrain that would allow a JSON `year : parameter : value` format?

**Comment by hdoupe** [comment]
I'm in favor of removing the python dictionary format. I think our primary concern should be choosing an order that is intuitive for the user. I think that the `parameter: year: value` format is more intuitive. That is, I think it's helpful, conceptually, to group the reform by parameter name before grouping by year. However, I haven't written a reform file that wasn't used purely for testing software since I started 2 years ago. So, I'd like to hear what others think.

I don't see any significant technical challenges to swapping to the `year: parameter: value` format. I would just have to swap a couple lines of code from:

```
revision[param][str(self.start_year + i)] = [parsed_input[param][i]]
```

to 

```
revision[str(self.start_year + i)][param] = [parsed_input[param][i]]
```

Tax-Calculator and the Tax-Brain python package are responsible for swapping from the `param: year: value` format to `year: param: value`. So, that's not something I have to worry about.

**Comment by hdoupe** [comment]
@martinholmer asked:

>  Are you willing and able to make the changes to TaxBrain that would allow a JSON year : parameter : value format?

Yes, I am. As I mentioned above, I don't see any technical difficulties. I think this could be a good opportunity to delete a bunch of code that does these conversions.

**Comment by martinholmer** [comment]
@hdoupe suggested in issue #2291:
> One option is to store the values for all years in the parameter's entry in the _vals dictionary:
> ...[snip]...
> Then, the active values are stored in the instance attribute. That way you don't have to fool with the underscore business.

Two things.  First, we **are** storing multiple-year data in`_vals`.  And even after PR #2292, each parameter in `_vals` has a leading underscore, which is important information to **developers** (but not to users) as @MattHJensen said earlier in this conversation.  The idea of getting rid of the leading underscores used internally in the code sounds simple, but the amount of code that would have to be changed is enormous and the distinction @MattHJensen pointed out would be lost.

The point of issue #2291 and pull request #2292 is not how to code Tax-Calculator internal logic, but rather it is what to require users to do in order to specify a policy reform.  Can we keep focused on that topic?

**Comment by hdoupe** [comment]
@martinholmer said:

> The point of issue #2291 and pull request #2292 is not how to code Tax-Calculator internal logic, but rather it is what to require users to do in order to specify a policy reform. Can we keep focused on that topic?

Yep, sorry about that.

**Comment by martinholmer** [comment]
@hdoupe said:
> Tax-Calculator and the Tax-Brain python package are responsible for swapping from the `param : year : value` format to `year : param : value`. So, that's not something I have to worry about.

OK.  So, then why does your [original comment](https://github.com/PSLmodels/Tax-Calculator/issues/2291#issuecomment-480947341) in issue #2291 dwell on this topic?

**Comment by hdoupe** [comment]
> OK. So, then why does your original comment in issue #2291 dwell on this topic?

You're right. I don't have to deal with converting between the two formats. However, I think the point still stands that there don't need to be two different parameter formats.

**Comment by martinholmer** [comment]
Originally, issue #2291 focused on a suggested revision to the syntax of **Python dictionaries** used to specify reforms like this one:
```
reform = {
   2020: {
       'SS_Earnings_c': 300000,
       'SS_Earnings_c-indexed': False
   },
   2021: {
       'STD': [14000, 28000, 14000, 21000, 28000]
   }
}
```
After useful discussion, the topic of issue #2291 has widened to include consideration of the best syntax for a JSON reform file.   Currently, PR #2292 leaves the JSON `parameter : year : value` format unchanged, but some want PR #2292 to change the JSON file format to be like the Python dictionary format.  If we did that, the JSON file representing the reform in the above Python dictionary would have this content:
```
{
  "policy": {
      "2020": {
         "SS_Earnings_c": 300000,
         "SS_Earnings_c-indexed": false
      },
      "2021": {
         "STD": [14000, 28000, 14000, 21000, 28000]
      }
}
```
So, what do people think about revising the required format of JSON reform files and assumption files?
I'm willing to add this change to PR #2292 if people think it is a good idea.

@MattHJensen @andersonfrailey @hdoupe @MaxGhenis @codykallen @feenberg

**Comment by martinholmer** [comment]
Turns out I missed an important point that @hdoupe made in the #2291 conversation.  Sorry about that.
Here is what he said recently:
> I'm in favor of removing the python dictionary format. I think our primary concern should be choosing an order that is intuitive for the user. I think that the parameter: year: value format is more intuitive. That is, I think it's helpful, conceptually, to group the reform by parameter name before grouping by year. However, I haven't written a reform file that wasn't used purely for testing software since I started 2 years ago. So, I'd like to hear what others think.

So, let me show in this comment what the Python dictionary (for the example reform in #2291) would look like if we followed @hdoupe's suggestion:
```
reform = {
   'SS_Earnings_c': {2020: 300000},
   'SS_Earnings_c-indexed': {2020: False},
   'STD': {2021: [14000, 28000, 14000, 21000, 28000]}
}
```
versus what it would look like if we did not change the parameter/year ordering:
```
reform = {
   2020: {
       'SS_Earnings_c': 300000,
       'SS_Earnings_c-indexed': False
   },
   2021: {
       'STD': [14000, 28000, 14000, 21000, 28000]
   }
}
```
Which should we go with in pull request #2292?

@MattHJensen @andersonfrailey @hdoupe @MaxGhenis @codykallen @feenberg

**Comment by andersonfrailey** [comment]
I personally prefer the `parameter:year:value` format for specifying reforms because if I'm writing a reform file and there is one parameter I adjust in multiple years, I like having all of the changes I'm making for that parameter to be together. For example:

```
{
  "ID_StateLocalTax_hc": {
          2020: 0.25,
          2021: 0.50,
          2022: 0.75,
          2023: 1.0
  }
}
```
Rather than:

```
{
  2020: {"ID_StateLocalTax_hc": 0.25},
  2021: {"ID_StateLocalTax_hc": 0.50},
  2022: {"ID_StateLocalTax_hc": 0.75},
  2023: {"ID_StateLocalTax_hc": 1.0}
}
```

 I don't see a huge difference in the level of intuitiveness of `parameter:year:value` vs. `year:parameter:value` as I can understand the logic behind both. My preferences just tend to the former. That being said, no matter which format we settle on I do think we should have one universal format for specifying policy reforms.

I'm also on board with @martinholmer's suggestion in #2292 to go ahead and merge that PR when it's ready and continue this discussion separately. I'm also happy to work with @hdoupe to modify TaxBrain as needed in response to any changes.

**Comment by martinholmer** [comment]
Issue #2291 has been resolved by the merge of pull request #2292.
